rustworkx-core/src/connectivity/core_number.rs

```diff
@@ -0,0 +1,206 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use crate::dictmap::*;
+use hashbrown::{HashMap, HashSet};
+use petgraph::visit::{GraphBase, IntoNeighborsDirected, IntoNodeIdentifiers, NodeCount};
+use petgraph::Direction::{Incoming, Outgoing};
+use std::hash::Hash;
+
+/// Return the core number for each node in the graph.
+///
+/// A k-core is a maximal subgraph that contains nodes of degree k or more.
+///
+/// The function implicitly assumes that there are no parallel edges
+/// or self loops. It may produce incorrect/unexpected results if the
+/// input graph has self loops or parallel edges.
+///
+/// Arguments:
+///
+/// * `graph` - The graph in which to find the core numbers.
+///
+/// # Example
+/// ```rust
+/// use petgraph::prelude::*;
+/// use rustworkx_core::connectivity::core_number;
+///
+/// let edge_list = vec![(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)];
+/// let graph = DiGraph::<i32, i32>::from_edges(&edge_list);
+/// let res: Vec<(usize, usize)> = core_number(graph)
+///     .iter()
+///     .map(|(k, v)| (k.index(), *v))
+///     .collect();
+/// assert_eq!(res, vec![(0, 3), (1, 3), (2, 3), (3, 3)]);
+/// ```
+pub fn core_number<G>(graph: G) -> DictMap<G::NodeId, usize>
+where
+    G: GraphBase + NodeCount,
+    for<'b> &'b G: GraphBase<NodeId = G::NodeId> + IntoNodeIdentifiers + IntoNeighborsDirected,
+    G::NodeId: Eq + Hash,
+{
+    let node_num = graph.node_count();
+    if node_num == 0 {
+        return DictMap::new();
+    }
+
+    let mut cores: DictMap<G::NodeId, usize> = DictMap::with_capacity(node_num);
+    let mut node_vec: Vec<G::NodeId> = graph.node_identifiers().collect();
+    let mut degree_map: HashMap<G::NodeId, usize> = HashMap::with_capacity(node_num);
+    let mut nbrs: HashMap<G::NodeId, HashSet<G::NodeId>> = HashMap::with_capacity(node_num);
+    let mut node_pos: HashMap<G::NodeId, usize> = HashMap::with_capacity(node_num);
+
+    for k in node_vec.iter() {
+        let k_nbrs: HashSet<G::NodeId> = graph
+            .neighbors_directed(*k, Incoming)
+            .chain(graph.neighbors_directed(*k, Outgoing))
+            .collect();
+        let k_deg = k_nbrs.len();
+
+        nbrs.insert(*k, k_nbrs);
+        cores.insert(*k, k_deg);
+        degree_map.insert(*k, k_deg);
+    }
+    node_vec.sort_by_key(|k| degree_map.get(k));
```

Comment:
Food for thought: should we remove the parallel sort? It might be worth benchmarking for large graphs

Reply:
No opinion on it. I'll try a quick benchmark and see if it shows a difference.

Reply:
Ran it with heavy hex with low core number and complete with high core number on my local and there was less that 5% variation between `sort_by_key` and `par_sort_by_key`. Maybe for some other types of graphs it could be different.

Reply:
I’d import Rayon and use `par_sort_by_key`. Rayon is in the dependencies already, might as well use it as the previous code for consistency

Reply:
Ok. I'll put `par_sort_by_key` back.